File: lib/domain/use_cases/find_communes_with_waiting.dart
import '../models/client.dart';
import 'client_status.dart';

typedef CommuneWithWaiting = ({String name, int count});

class FindCommunesWithWaiting {
  const FindCommunesWithWaiting();

  List<CommuneWithWaiting> call({
    required List<Client> clients,
    required Map<int, ClientStatus> statusByClientId,
  }) {
    final counts = <String, int>{};
    for (final c in clients) {
      if (c.needsDistanceRecompute) continue;
      if (statusByClientId[c.id] != ClientStatus.waiting) continue;
      counts[c.city] = (counts[c.city] ?? 0) + 1;
    }
    final entries = counts.entries
        .map((e) => (name: e.key, count: e.value))
        .toList()
      ..sort((a, b) => a.name.compareTo(b.name));
    return entries;
  }
}

File: src/domain/use-cases/find-communes-with-waiting.ts
interface ClientLite {
  id: string;
  isWaiting: boolean;
  addressCity: string | null;
}

export interface CommuneCount {
  city: string;
  count: number;
}

export function findCommunesWithWaiting(clients: ClientLite[]): CommuneCount[] {
  const counts = new Map<string, number>();
  for (const c of clients) {
    if (!c.isWaiting || !c.addressCity) continue;
    counts.set(c.addressCity, (counts.get(c.addressCity) ?? 0) + 1);
  }
  return [...counts.entries()]
    .map(([city, count]) => ({ city, count }))
    .sort((a, b) => b.count - a.count || a.city.localeCompare(b.city, 'fr'));
}
